pkg/controller/util/secret_test.go

```diff
@@ -0,0 +1,136 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestEnsureImageStreamTag(t *testing.T) {
+	ctx := context.Background()
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:   "ci",
+			Name:        "registry-pull-credentials",
+			Annotations: map[string]string{"a": "c"},
+		},
+		Data: map[string][]byte{"pass": []byte("some")},
+		Type: corev1.SecretTypeDockerConfigJson,
+	}
+
+	targetSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:   "some-ns",
+			Name:        "registry-pull-credentials",
+			Annotations: map[string]string{"a": "b"},
+		},
+		Data: map[string][]byte{"pass": []byte("other")},
+	}
+
+	testCases := []struct {
+		name      string
+		client    ctrlruntimeclient.Client
+		secret    *corev1.Secret
+		namespace string
+		expected  error
+		verify    func(client ctrlruntimeclient.Client) error
+	}{
+		{
+			name:      "basic case: create",
+			client:    fakeclient.NewFakeClient(secret.DeepCopy()),
+			namespace: "some-ns",
+			verify: func(client ctrlruntimeclient.Client) error {
+				actualSecret := &corev1.Secret{}
+				if err := client.Get(ctx, types.NamespacedName{Name: "registry-pull-credentials", Namespace: "some-ns"}, actualSecret); err != nil {
+					return err
+				}
+				expectedSecret := &corev1.Secret{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Secret",
+						APIVersion: "v1",
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace:       "some-ns",
+						Name:            "registry-pull-credentials",
+						Annotations:     map[string]string{"a": "c"},
+						ResourceVersion: "1",
+					},
+					Data: map[string][]byte{"pass": []byte("some")},
+					Type: "kubernetes.io/dockerconfigjson",
+				}
+
+				if diff := cmp.Diff(expectedSecret, actualSecret); diff != "" {
+					return fmt.Errorf("actual does not match expected, diff: %s", diff)
+				}
+				return nil
+			},
+		},
+		{
+			name:      "basic case: update",
+			client:    fakeclient.NewFakeClient(secret.DeepCopy(), targetSecret.DeepCopy()),
+			namespace: "some-ns",
+			verify: func(client ctrlruntimeclient.Client) error {
+				actualSecret := &corev1.Secret{}
+				if err := client.Get(ctx, types.NamespacedName{Name: "registry-pull-credentials", Namespace: "some-ns"}, actualSecret); err != nil {
+					return err
+				}
+				expectedSecret := &corev1.Secret{
+					TypeMeta: metav1.TypeMeta{
```

Comment:
Same comment regarding the comparer func to make this less verbose

Reply:
Updated.